File: SexyBiscuit.Engine/Core/ExceptionIsolation.cs
namespace SexyBiscuit.Engine.Core;

/// <summary>
/// Lets a host survive game code that throws. With no handler installed exceptions propagate
/// exactly as before, which is what a shipped game wants: fail loudly. The editor installs a
/// handler that logs the exception, disables a component that keeps throwing, and keeps the
/// frame loop alive.
/// </summary>
public static class ExceptionIsolation
{
    /// <summary>
    /// Called with the exception, the actor or component that threw, and the lifecycle phase.
    /// Return true to swallow the exception, false to let it propagate.
    /// </summary>
    public static Func<Exception, object, string, bool>? Handler { get; set; }

    /// <summary>Used as an exception filter: true only when a handler exists and claims the exception.</summary>
    public static bool TryHandle(Exception exception, object owner, string phase)
    {
        var handler = Handler;
        if (handler == null) return false;

        try
        {
            return handler(exception, owner, phase);
        }
        catch (Exception)
        {
            return false;
        }
    }
}

File: html5/src/core/ExceptionIsolation.js
// -----------------------------------------------------------------------------
// ExceptionIsolation — one component throwing must not take down the editor.
// -----------------------------------------------------------------------------

let _handler = null;

/**
 * With no handler installed, `tryHandle` returns false and the caller rethrows,
 * so a shipped game keeps its fail-fast behaviour. The editor installs one so a
 * throwing game component logs to the console panel and is skipped instead.
 */
export const ExceptionIsolation = {
    /**
     * @param {?(error: any, owner: object, phase: string) => boolean} handler
     *   Returns true to swallow the error, false to let it propagate.
     */
    setHandler(handler) { _handler = handler; },

    get handler() { return _handler; },

    /** True when the error was handled and the caller should continue. */
    tryHandle(error, owner, phase) {
        if (!_handler) return false;
        try {
            return _handler(error, owner, phase) === true;
        } catch {
            // A throwing handler is worse than no handler; fall back to propagating.
            return false;
        }
    },
};

/**
 * Runs `fn` with the isolation policy applied. Used by every lifecycle
 * call site so the try/catch shape lives in exactly one place.
 */
export function isolate(fn, owner, phase) {
    try {
        return fn();
    } catch (error) {
        if (!ExceptionIsolation.tryHandle(error, owner, phase)) throw error;
        return undefined;
    }
}
